src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/PersonService.kt

```diff
@@ -102,16 +102,25 @@ class PersonService(
   private fun isCreateEvent(event: String?) = listOf(PRISONER_CREATED, NEW_OFFENDER_CREATED).contains(event)
 
   fun findCandidateRecords(person: Person): Page<PersonEntity> {
-    val postcodeSpecifications = person.addresses.map { PersonSpecification.levenshteinPostcode(it.postcode) }
+    val postcodeSpecifications = person.addresses.map {
+      PersonSpecification.isNotNull(it.postcode).and(PersonSpecification.levenshteinPostcode(it.postcode))
+    }
 
     val soundexFirstLastName = Specification.where(
-      PersonSpecification.soundex(person.firstName, PersonSpecification.FIRST_NAME)
-        .and(PersonSpecification.soundex(person.lastName, PersonSpecification.LAST_NAME)),
+      PersonSpecification.isNotNull(person.firstName).and(PersonSpecification.isNotNull(person.lastName))
+        .and(
+          PersonSpecification.soundex(person.firstName, PersonSpecification.FIRST_NAME)
+            .and(PersonSpecification.soundex(person.lastName, PersonSpecification.LAST_NAME)),
+        ),
+    )
+
+    val levenshteinDob = Specification.where(
+      PersonSpecification.isNotNull(person.dateOfBirth.toString())
```

Comment:
date of birth is nullable, so calling toString on it when it is null will return the string `null`. Have a look at this in a scratch file

`import java.time.LocalDate

val x:LocalDate? = null
val dateToString = x.toString()
println(x)
println(dateToString == null)`

Reply:
Moved the null check into criteria builder